packages/runtime-core/src/component.ts

```diff
@@ -62,6 +62,40 @@ export type Data = Record<string, unknown>
  */
 export interface ComponentCustomProps {}
 
+/**
+ * For globally defined Directives
+ * Here is an example of adding a directive `VTooltip` as global directive:
+ *
+ * @example
+ * ```ts
+ * import VTooltip from 'v-tooltip'
+ *
+ * declare module '@vue/runtime-core' {
+ *   interface GlobalDirectives {
+ *     VTooltip
+ *   }
+ * }
+ * ```
+ */
+export interface GlobalDirectives extends Record<string, Directive> {}
+
+/**
+ * For globally defined Components
+ * Here is an example of adding a component `RouterView` as global component:
+ *
+ * @example
+ * ```ts
+ * import { RouterView } from 'vue-router'
+ *
+ * declare module '@vue/runtime-core' {
+ *   interface GlobalComponents {
+ *     RouterView
+ *   }
+ * }
+ * ```
+ */
+export interface GlobalComponents extends Record<string, Component> {}
```

Comment:
I think `Transition`, `TransitionGroup`, `KeepAlive`, `Suspense`, `Teleport` should be include to here.

Reply:
Added all the Components and Directives (only `vShow`) as Global

Reply:
@pikax `GlobalComponents` extends `Record<string, Component>`, so can we infer if a component is global or not?

I am using following type helper to resolve component (for type check only):

```ts
function resolveComponent<T extends {}, A, B>(
  localRegisteredComponents: T,
  tagName: A,
  tagNameInPascalName?: B,
): A extends keyof T
  ? T[A]
  : B extends keyof T
  ? T[B]
  : A extends keyof KnownKeys<JSX.IntrinsicElements>
  ? ComponentFromProps<JSX.IntrinsicElements[A]>
  : B extends keyof KnownKeys<JSX.IntrinsicElements>
  ? ComponentFromProps<JSX.IntrinsicElements[B]>
  : A extends keyof GlobalComponents
  ? GlobalComponents[A]
  : B extends keyof GlobalComponents
  ? GlobalComponents[B]
  : never
  
  
type ComponentFromProps<P> = new () => {
  $props: P
  $slots: { default(): VNodeChild }
}
```
It should return `never` when component is not registered locally, not an intrinsic element and not registered globally.

Reply:
@znck interesting, the reason why it extends `Record<string, Component>` is to allow typecheck when users override it, what would be your use case if the Component is global?